numba/parfor.py

```diff
@@ -498,88 +825,256 @@ def _convert_prange(self, blocks):
+                        """find expressions that involve getitem using the
+                        index variable. Return both the arrays and expressions.
+                        """
+                        indices = copy.copy(loop_index_vars)
+                        for block in loop_body.values():
+                            for inst in block.find_insts(ir.Assign):
+                                if (isinstance(inst.value, ir.Var) and
+                                    inst.value.name in indices):
+                                    indices.add(inst.target.name)
+                        arrs = []
+                        exprs = []
+                        for block in loop_body.values():
+                            for inst in block.body:
+                                lv = set(x.name for x in inst.list_vars())
+                                if lv & indices:
+                                    if lv.issubset(indices):
+                                        continue
+                                    require(isinstance(inst, ir.Assign))
+                                    expr = inst.value
+                                    require(isinstance(expr, ir.Expr) and
+                                            expr.op == 'getitem')
+                                    arrs.append(expr.value.name)
+                                    exprs.append(expr)
+                        return arrs, exprs
+
+                    mask_var = None
+                    mask_indices = None
+                    def find_mask_from_size(size_var):
+                        """Find the case where size_var is defined by A[M].shape,
+                        where M is a boolean array.
+                        """
+                        size_def = get_definition(self.func_ir, size_var)
+                        require(size_def and isinstance(size_def, ir.Expr) and
+                                size_def.op == 'getattr' and size_def.attr == 'shape')
+                        arr_var = size_def.value
+                        live_vars = set.union(*[live_map[l] for l in loop.exits])
+                        index_arrs, index_exprs = find_indexed_arrays()
+                        require([arr_var.name] == list(index_arrs))
+                        # input array has to be dead after loop
+                        require(arr_var.name not in live_vars)
+                        # loop for arr's definition, where size = arr.shape
+                        arr_def = get_definition(self.func_ir, size_def.value)
+                        result = self._find_mask(arr_def)
+                        # Found the mask.
+                        # Replace B[i] with A[i], where B = A[M]
+                        for expr in index_exprs:
+                            expr.value = result[0]
+                        return result
+
+                    # TODO: support array mask optimization for prange
+                    if loop_kind == 'pndindex':
```

Comment:
I added some comments but @ninegua knows this code better. I think this piece of code needs major refactoring so that non-functional programmers like me can understand it better :)